R/checkout.R

```diff
@@ -2,21 +2,48 @@
 #' Checkout a Repository
 #'
 #' `renv::checkout()` can be used to install the latest packages available from
-#' the requested repositories.
+#' a (set of) package repositories.
+#'
+#' `renv::checkout()` is most useful with services like the Posit's
+#' [Package Manager](https://packagemanager.rstudio.com/), as it
+#' can be used to switch between different repository snapshots within an
+#' `renv` project. In this way, you can upgrade (or downgrade) all of the
+#' packages used in a particular `renv` project to the package versions
+#' provided by a particular snapshot.
 #'
 #' @inheritParams renv-params
 #'
-#' @param repos The \R package repositories to check out. When `NULL` (the default),
-#'   `renv` will attempt to determine the packages used in the project via
-#'   a call to [renv::dependencies()].
+#' @param repos The \R package repositories to check out.
 #'
 #' @param packages The packages to be installed. When `NULL` (the default),
-#'   all packages currently used in the project will be installed.
+#'   all packages currently used in the project will be installed, as
+#'   determined by [renv::dependencies()]. The recursive dependencies of these
+#'   packages will be included as well.
+#'
+#' @param date The snapshot date to use. When set, the associated snapshot as
```

Comment:
I think ideally we'd recognise the an existing repo named `RPSM` either in `getOptions("repos")` or the `repos` argument.

Reply:
Sorry, I don't understand -- can you elaborate?

Reply:
FWIW `date` is really just a shortcut so you can just use `renv::checkout(date = "2022-01-02")` instead of `renv::checkout(repos = "https://packagemanager.rstudio.com/cran/2022-01-02")` or something similar.

Reply:
Maybe there are really two commands here — set up my repos (with convenience syntax for setting dates) and re-install all packages from those repos (reporting where you can't).

Reply:
The potential for users to have multiple PPM instances in their repositories makes me think we should do away with the `date` parameter, and ask users to just provide the full repository URL. This lets users take charge of what copy of PPM they're using, or whether they need a "mix" of snapshots from two different PPM instances for some reason.

The downside is (if you have scripts using `renv::checkout()`), if you need to change the repository URL in the future, you need to change all of those `renv::checkout()` calls to use the updated repository. Not sure if that's a real concern.

Otherwise, we might want to have something like:

- A project setting giving the "root" URL for the PPM instance,
- (Maybe?) A project setting giving the repository name to be used from the PPM instance (e.g. "cran"),
- `renv::checkout(date = <...>)` then makes use of the project settings when forming the full repository URL.

Or maybe (similar to Hadley's suggestion) we have a project setting giving a map from repository names to PPM "root" URLs, e.g.

```
{
    "RSPM": "https://pm.company.local/cran",
    "LOCAL": "https://pm.company.local/local"
}
```

and then the requested date would be applied to all "known" PPM URLs that have been registered in called to `renv::checkout(date = <...>)`. The only requirement then would be that the names of the `getOption("repos")` option would match the names in the project setting.

Reply:
Some folks might have:

```r
options(repos = c(
    CRAN = "https://pm-dev.company.local/cran/2023-01-01"
))
```

and then when they deploy their content into some production environment, they'll want us to use:

```r
options(repos = c(
    CRAN = "https://pm-prod.company.com/cran/XXX"
))
```

This transition is not the same as "checkout", though. In this situation, folks want to use the same package graph but obtain those packages from a different repository URL.

Package Manager snapshot URLs (either date-stamps or transaction ids) are not valid across servers (except in the very narrow case where the repository is a simple, direct mirror of a single source like CRAN). Many installations use "approved CRAN" configurations. Rewriting from "dev" to "prod" essentially means that you need to drop the snapshot reference, and then you're asking for package binary incompatibility..

I really like the idea of "update my library to use 'current' versions" for whatever is the active set of repositories.

Reply:
Aside: We should discourage `RSPM="https://pm.company.local/cran"` because `RSPM` is the server identifier included by RSPM in the `Repository: RSPM` package `DESCRIPTION` entry. It is unfortunately identifying the kind of the server, not the name of the repository.

Reply:
Coming all the way back around, maybe it's still useful to have `date` create a repo using PPPM — that makes it easy for folks who can use PPPM, while not making it any harder for the folks with internal repos that need to supply the full url.